### PR TITLE
feat: Exit while cache is not configured correctly

### PR DESCRIPTION
**Pull request by Xuanwo**
Signed-off-by: Xuanwo <github@xuanwo.io>

This PR is the first step of https://github.com/mozilla/sccache/issues/1512.

We will exit while the cache is not configured correctly. For example, s3 is configured with an invalid endpoint.

### CODE REVIEW

File: tests/sccache_cargo.rs [hunk 1]
```diff
@@ -302,24 +302,11 @@ fn test_gcp_arg_check() -> Result<()> {
...
     cmd.arg("--start-server")
         .env("SCCACHE_LOG", "debug")
-        .env("SCCACHE_GCS_OAUTH_URL", "http://foobar");
+        .env("SCCACHE_GCS_OAUTH_URL", "http://127.0.0.1");
 
     cmd.assert().failure().stderr(predicate::str::contains(
         "If setting GCS credentials, SCCACHE_GCS_BUCKET",
     ));
 
-    stop_sccache()?;
```
Comment (sylvestre) [inline tests/sccache_cargo.rs:311]: why did you remove it?

Comment (sylvestre) [reply tests/sccache_cargo.rs:311]: maybe we should keep/transform it to have an invalid endpoint to test the error ?

Comment (Xuanwo) [reply tests/sccache_cargo.rs:311]: Nice catch. I found current logic may break the server spawn logic. Trying to fix.

**Comment by Xuanwo** [comment]
Fixed! @sylvestre PTAL.

File: tests/sccache_cargo.rs [hunk 2]
```diff
@@ -313,13 +313,13 @@ fn test_gcp_arg_check() -> Result<()> {
...
+        .env("SCCACHE_GCS_CREDENTIALS_URL", "not_valid_url//127.0.0.1")
         .env("SCCACHE_GCS_KEY_PATH", "foo.json");
 
     // This is just a warning
-    cmd.assert().success().stderr(predicate::str::contains(
-        "SCCACHE_GCS_OAUTH_URL has been deprecated",
-    ));
+    cmd.assert()
+        .failure()
+        .stderr(predicate::str::contains("gcs credential url is invliad"));
```
Comment (sylvestre) [inline tests/sccache_cargo.rs:322]:
```suggestion
        .stderr(predicate::str::contains("gcs credential url is invalid"));
```

File: src/cache/cache.rs [hunk 3]
```diff
@@ -373,30 +382,31 @@ impl Storage for opendal::Operator {
 }
 
 /// Normalize key `abcdef` into `a/b/c/abcdef`
+#[allow(dead_code)]
```
Comment (drahnr) [inline src/cache/cache.rs:385]: Could it be removed altogether?

File: src/cache/cache.rs [hunk 4]
```diff
@@ -405,99 +415,87 @@ pub fn storage_from_config(config: &Config, pool: &tokio::runtime::Handle) -> Ar
...
             }
+            #[cfg(not(any(
+                feature = "azure",
+                feature = "gcs",
+                feature = "gha",
+                feature = "memcached",
+                feature = "redis",
+                feature = "s3"
+            )))]
+            _ => return Err(anyhow!("cache type is not enabled")),
```
Comment (drahnr) [inline src/cache/cache.rs:492]:
```suggestion
            _ => bail!("cache type is not enabled"),
```

File: src/config.rs [hunk 5]
```diff
@@ -666,7 +665,7 @@ fn config_file(env_var: &str, leaf: &str) -> PathBuf {
 
 #[derive(Debug, Default, PartialEq, Eq)]
 pub struct Config {
-    pub caches: Vec<CacheType>,
+    pub cache: Option<CacheType>,
```
Comment (drahnr) [inline src/config.rs:668]: Was this used anywhere before? Did multiple backends ever work?

**Review by drahnr — APPROVED** [review]
It's unclear whether the changes of many backends to one is diserable or not.

File: src/config.rs (hunk 5 above)
Comment (Xuanwo) [reply src/config.rs:668]: > Did multiple backends ever work?

No, they never works at the same time.

The only case is users configure different backends at the same time, and some of them are wrong. For example, users configure s3, azure, redis at the same time, and s3, azure is invalid, only redis works.

In old logic, the errors returned by s3 and azure will be ignore, and using redis finally. In this PR, we will only choose the first valid cache type and returns error happened which makes the behavior more predictable.

In the future, I will return errors if users configured different backends at the same time.

File: src/cache/cache.rs (hunk 3 above)
Comment (Xuanwo) [reply src/cache/cache.rs:385]: This function will be used if feature like `s3`, `azure` enabled. Adding a `dead_code` is the most simple way to address clippy errors under `--no-default-feature`.
